### ISSUE TITLE
Notebook tutorial fails to run

### ISSUE THREAD

**Issue by kgreggy** [durham-qlm/scope2elecsus#1]
When I attempted to use the tutorial, I restarted and ran all cells in the kernal. In the data fitting section starting at cell 63, I got the error: 'Missing independent variable 'verbose'. 

The error cell says this issue occurs in line 20 of this cell, which is failing in line 181 of elecsus_methods.py, pointing to line 148 in MLFittingRountine.py. It appears to be failing to pass verbose defined in elecsus_methods.py all the way down to the lmfit code.

lmfit is updated to the 1.3.1 version on my machine.

**Comment by xsection** [comment]
Hi,

I asked a graduate student to look into this. He uses Elecsus regularly. He could not reproduce the error. Could you supply any more information or send screenshots etc.?

Thanks,

Steven

------------------------------------------------------
Dr. Steven A. Wrathmall
Department of Physics
Durham University
Science Laboratories
South Road
Durham DH1 3LE
UK
Staff page: https://www.durham.ac.uk/staff/s-a-wrathmall/
+44 (0) 191 334 3763 (office)
+44 (0) 191 334 5823 (fax)

Pronouns: he/him/his

Due to my own work/life balance, you may get emails from me outside of normal working hours. PLEASE DO NOT respond outside of your own working pattern.
------------------------------------------------------


From: kgreggy ***@***.***>
Date: Thursday, 9 May 2024 at 16:56
To: durham-qlm/scope2elecsus ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [durham-qlm/scope2elecsus] Notebook tutorial fails to run (Issue #1)
[EXTERNAL EMAIL]

When I attempted to use the tutorial, I restarted and ran all cells in the kernal. In the data fitting section starting at cell 63, I got the error: 'Missing independent variable 'verbose'.

The error cell says this issue occurs in line 20 of this cell, which is failing in line 181 of elecsus_methods.py, pointing to line 148 in MLFittingRountine.py. It appears to be failing to pass verbose defined in elecsus_methods.py all the way down to the lmfit code.

lmfit is updated to the 1.3.1 version on my machine.

**Comment by kgreggy** [comment]
Hi Steven,

I am running anaconda from MacOS Sonoma 14.4.1; I updated all packages and checked the relevant ones were installed yesterday after first getting the error seen in the attachment. The python version is 3.8.19 and the notebook version is listed as "Version 29 07 2022" at the very bottom. I tried changing the file that calls ML.ML_fit() to include verbose and re-running the kernal, but that did not solve the issue.

 Is your student running LMFit 1.3.1? It came out a few weeks ago, and may be causing new issues.

Thanks,
Kenny
________________________________

**Comment by ghost** [comment]
Hi,

I've had a look at this issue and I can't reproduce the error. My version of lmfit is 1.0.3 (older than I thought!). 

Line 148 in my version of MLFittingRountine.py (part of the ML_fit function) is: if verbose: print(params). Have you tried removing this line or commenting it out just to see if that solves the issue?

Thanks,
Jack

**Comment by kgreggy** [comment]
Hi Jack,

line 148 for you is line 146 for me; the line I am having issues with is: "result = model.fit(y,x=x, params=params, method=method, **kwargz))".

When I save changes to the file and restart my kernal, it still uses the original version without the changes I have made. I have tried setting verbose directly to True/False, along with verbose=verbose.

Am I missing something about importing libraries? I thought that if I save the changes to an internal file and re-import the new file would be loaded in.

I am planning on having another student in my group try what I am doing, to see if the issue is only on my end.

Thanks,
Kenny

________________________________

**Comment by jbriscoeDUphysics** [comment]
Hi Kenny,

That's interesting. Perhaps there is a clash with the new version of lmfit and any implementation of verbose, but I'm not sure. I have written some code which you could try. I added this to the Notebook Tutorial in a new cell above the markdown cell: "In the next three cells, we fit the three spectra using ElecSus. More details on the program can be found in the ElecSus documentation. ". This is just some fitting code which calls the ElecSus function get_spectra, and fits using lmfit, without calling the MLFittingRountine.py module. This can be modified for more complicated fits.

from elecsus.libs.spectra import get_spectra as gs
import lmfit as lm

def fit_spectra(x, T, p_dict={}):    
    '''A function which uses ElecSus and input linear horizontal light to fit T to an absorption spectrum'''

    p_dict['T'] = T # T may already be in the dictionary, but this line allow the fitting algorithm to vary it.  

    E_in = np.array([1, 0, 0])
        
    [tau] = gs(x, E_in, p_dict, outputs=['S0'])
        
    return tau.real

fit_kwarg_dict = {'Elem':'Rb','Dline':'D2','lcell':75e-3,'GammaBuf':0, 'Bphi': (0/90)*np.pi/2, # known parameters.
                  'rb85frac':72.17, 'Btheta':0*np.pi/180}
fit_params = {'T':True}  # vary temperature.

# any values where the contrain range is really small ensures this parameter remains constant. They could be fit.
fit_dict = {'T':20} # pick an approximate starting temperature for the fit.
fit_bounds = {'T':(0, 100)} # sets bounds of temperature. This will need to be changed for different datasets.

fitFn = fit_spectra # function used to fit spectra.
fitModel = lm.Model(fitFn) # call lmfit.

fitParams = fitModel.make_params(**fit_dict)
kwargs = {'p_dict':fit_kwarg_dict}

allkeys = fitParams.valuesdict()
for key in allkeys:
    fitParams[key].vary = False
    
for key in fit_params:
    fitParams[key].vary = fit_params[key]
    
    if key in fit_params:
        fitParams[key].min = fit_bounds[key][0]
        fitParams[key].max = fit_bounds[key][1]
        
ydata = Vals_40 # set to Vals_20, Vals_40, or Vals_73. You may need  to adjust fit bounds above accordingly.
xdata = Time_40 # set to Time_20, Time_40, or Time_73.

fitResult = fitModel.fit(ydata, x = xdata,
                         params=fitParams,method='differential_evolution',**kwargs)
print(fitResult.fit_report())

fitCurve = fitResult.best_fit
fitEndParams = fitResult.best_values

RMSE = np.sqrt(np.mean((ydata - fitCurve)**2))
print('Fit RMS Error (%):\t',100*RMSE)

fig, ax = plt.subplots(2,sharex=True)
ax[0].plot(xdata,ydata,'k.',label='Data')
ax[0].plot(xdata,fitCurve,'r-',lw=2,label='Fit')
ax[0].set_ylabel('Transmission (S0)')
ax[1].set_xlabel('Linear Detuning (GHz)')
#Residuals x100
ax[1].scatter(xdata,100*(fitCurve-((ydata))))

**Comment by jbriscoeDUphysics** [comment]
Hi Kenny,

That's interesting. Perhaps there is a clash with the new version of lmfit and any implementation of verbose, but I'm not sure. I have written some code which you could try. I added this to the Notebook Tutorial in a new cell above the markdown cell: "In the next three cells, we fit the three spectra using ElecSus. More details on the program can be found in the ElecSus documentation. ". This is just some fitting code which calls the ElecSus function get_spectra, and fits using lmfit, without calling the MLFittingRountine.py module. This can be modified for more complicated fits.
`
from elecsus.libs.spectra import get_spectra as gs
import lmfit as lm

def fit_spectra(x, T, p_dict={}):    
    '''A function which uses ElecSus and input linear horizontal light to fit T to an absorption spectrum'''

    p_dict['T'] = T # T may already be in the dictionary, but this line allow the fitting algorithm to vary it.  

    E_in = np.array([1, 0, 0])
        
    [tau] = gs(x, E_in, p_dict, outputs=['S0'])
        
    return tau.real

fit_kwarg_dict = {'Elem':'Rb','Dline':'D2','lcell':75e-3,'GammaBuf':0, 'Bphi': (0/90)*np.pi/2, # known parameters.
                  'rb85frac':72.17, 'Btheta':0*np.pi/180}
fit_params = {'T':True}  # vary temperature.

# any values where the contrain range is really small ensures this parameter remains constant. They could be fit.
fit_dict = {'T':20} # pick an approximate starting temperature for the fit.
fit_bounds = {'T':(0, 100)} # sets bounds of temperature. This will need to be changed for different datasets.

fitFn = fit_spectra # function used to fit spectra.
fitModel = lm.Model(fitFn) # call lmfit.

fitParams = fitModel.make_params(**fit_dict)
kwargs = {'p_dict':fit_kwarg_dict}

allkeys = fitParams.valuesdict()
for key in allkeys:
    fitParams[key].vary = False
    
for key in fit_params:
    fitParams[key].vary = fit_params[key]
    
    if key in fit_params:
        fitParams[key].min = fit_bounds[key][0]
        fitParams[key].max = fit_bounds[key][1]
        
ydata = Vals_40 # set to Vals_20, Vals_40, or Vals_73. You may need  to adjust fit bounds above accordingly.
xdata = Time_40 # set to Time_20, Time_40, or Time_73.

fitResult = fitModel.fit(ydata, x = xdata,
                         params=fitParams,method='differential_evolution',**kwargs)
print(fitResult.fit_report())

fitCurve = fitResult.best_fit
fitEndParams = fitResult.best_values

RMSE = np.sqrt(np.mean((ydata - fitCurve)**2))
print('Fit RMS Error (%):\t',100*RMSE)

fig, ax = plt.subplots(2,sharex=True)
ax[0].plot(xdata,ydata,'k.',label='Data')
ax[0].plot(xdata,fitCurve,'r-',lw=2,label='Fit')
ax[0].set_ylabel('Transmission (S0)')
ax[1].set_xlabel('Linear Detuning (GHz)')
#Residuals x100
ax[1].scatter(xdata,100*(fitCurve-((ydata))))
`

**Comment by jbriscoeDUphysics** [comment]
Hi Kenny,

That's interesting. Perhaps there is a clash with the new version of lmfit and any implementation of verbose, but I'm not sure. I have written some code which you could try. I added this to the Notebook Tutorial in a new cell above the markdown cell: "In the next three cells, we fit the three spectra using ElecSus. More details on the program can be found in the ElecSus documentation. ". This is just some fitting code which calls the ElecSus function get_spectra, and fits using lmfit, without calling the MLFittingRountine.py module. This can be modified for more complicated fits.

``
from elecsus.libs.spectra import get_spectra as gs
import lmfit as lm

def fit_spectra(x, T, p_dict={}):    
    '''A function which uses ElecSus and input linear horizontal light to fit T to an absorption spectrum'''

    p_dict['T'] = T # T may already be in the dictionary, but this line allow the fitting algorithm to vary it.  

    E_in = np.array([1, 0, 0])
        
    [tau] = gs(x, E_in, p_dict, outputs=['S0'])
        
    return tau.real

fit_kwarg_dict = {'Elem':'Rb','Dline':'D2','lcell':75e-3,'GammaBuf':0, 'Bphi': (0/90)*np.pi/2, # known parameters.
                  'rb85frac':72.17, 'Btheta':0*np.pi/180}
fit_params = {'T':True}  # vary temperature.

# any values where the contrain range is really small ensures this parameter remains constant. They could be fit.
fit_dict = {'T':20} # pick an approximate starting temperature for the fit.
fit_bounds = {'T':(0, 100)} # sets bounds of temperature. This will need to be changed for different datasets.

fitFn = fit_spectra # function used to fit spectra.
fitModel = lm.Model(fitFn) # call lmfit.

fitParams = fitModel.make_params(**fit_dict)
kwargs = {'p_dict':fit_kwarg_dict}

allkeys = fitParams.valuesdict()
for key in allkeys:
    fitParams[key].vary = False
    
for key in fit_params:
    fitParams[key].vary = fit_params[key]
    
    if key in fit_params:
        fitParams[key].min = fit_bounds[key][0]
        fitParams[key].max = fit_bounds[key][1]
        
ydata = Vals_40 # set to Vals_20, Vals_40, or Vals_73. You may need  to adjust fit bounds above accordingly.
xdata = Time_40 # set to Time_20, Time_40, or Time_73.

fitResult = fitModel.fit(ydata, x = xdata,
                         params=fitParams,method='differential_evolution',**kwargs)
print(fitResult.fit_report())

fitCurve = fitResult.best_fit
fitEndParams = fitResult.best_values

RMSE = np.sqrt(np.mean((ydata - fitCurve)**2))
print('Fit RMS Error (%):\t',100*RMSE)

fig, ax = plt.subplots(2,sharex=True)
ax[0].plot(xdata,ydata,'k.',label='Data')
ax[0].plot(xdata,fitCurve,'r-',lw=2,label='Fit')
ax[0].set_ylabel('Transmission (S0)')
ax[1].set_xlabel('Linear Detuning (GHz)')
#Residuals x100
ax[1].scatter(xdata,100*(fitCurve-((ydata))))
``

**Comment by ghost** [comment]
Hi Kenny,

That's interesting. Perhaps there is a clash with the new version of lmfit and any implementation of verbose, but I'm not sure. I would also expect if you make changes to the ElecSus files, save them, and restart the kernel, the new file should import.

I understand that's not too helpful so I have written some code which you could try. I added this to the Notebook Tutorial in a new cell above the markdown cell: "In the next three cells, we fit the three spectra using ElecSus. More details on the program can be found in the ElecSus documentation. ". This is just some fitting code which calls the ElecSus function get_spectra, and fits using lmfit, without calling the MLFittingRountine.py module. This can be modified for more complicated fits.

There is no use of verbose in this fitting method so hopefully it would work.

```
from elecsus.libs.spectra import get_spectra as gs
import lmfit as lm

def fit_spectra(x, T, p_dict={}):    
    '''A function which uses ElecSus and input linear horizontal light to fit T to an absorption spectrum'''

    p_dict['T'] = T # T may already be in the dictionary, but this line allows the fitting algorithm to vary it.  

    E_in = np.array([1, 0, 0])
        
    [tau] = gs(x, E_in, p_dict, outputs=['S0'])
        
    return tau.real

fit_kwarg_dict = {'Elem':'Rb','Dline':'D2','lcell':75e-3,'GammaBuf':0, 'Bphi': (0/90)*np.pi/2, # known parameters.
                  'rb85frac':72.17, 'Btheta':0*np.pi/180}
fit_params = {'T':True}  # vary temperature.

fit_dict = {'T':20} # pick an approximate starting temperature for the fit.
fit_bounds = {'T':(0, 100)} # sets bounds of temperature. This may need to be changed for different datasets.

fitFn = fit_spectra # function used to fit spectra.
fitModel = lm.Model(fitFn) # call lmfit.

fitParams = fitModel.make_params(**fit_dict)
kwargs = {'p_dict':fit_kwarg_dict}

allkeys = fitParams.valuesdict()
for key in allkeys:
    fitParams[key].vary = False
    
for key in fit_params:
    fitParams[key].vary = fit_params[key]
    
    if key in fit_params:
        fitParams[key].min = fit_bounds[key][0]
        fitParams[key].max = fit_bounds[key][1]
        
ydata = Vals_40 # set to Vals_20, Vals_40, or Vals_73. You may need  to adjust fit bounds above accordingly.
xdata = Time_40 # set to Time_20, Time_40, or Time_73.

fitResult = fitModel.fit(ydata, x = xdata,
                         params=fitParams,method='differential_evolution',**kwargs)
print(fitResult.fit_report())

fitCurve = fitResult.best_fit
fitEndParams = fitResult.best_values

RMSE = np.sqrt(np.mean((ydata - fitCurve)**2))
print('Fit RMS Error (%):\t',100*RMSE)

fig, ax = plt.subplots(2,sharex=True)
ax[0].plot(xdata,ydata,'k.',label='Data')
ax[0].plot(xdata,fitCurve,'r-',lw=2,label='Fit')
ax[0].set_ylabel('Transmission (S0)')
ax[1].set_xlabel('Linear Detuning (GHz)')
#Residuals x100
ax[1].scatter(xdata,100*(fitCurve-((ydata))))
```

Thanks,
Jack
